[KHR] Provide an alternative to <sycl/sycl.hpp>

AlexeySachkov opened the pull request:
This PR aims to address KhronosGroup/SYCL-Docs#780 by introducing a new KHR extension that document more headers that SYCL applications can include for better control of what they pay for in terms of compile time.

AlexeySachkov committed:
[KHR] Provide an alternative to <sycl/sycl.hpp>

This PR aims to address KhronosGroup/SYCL-Docs#780 by introducing a new
KHR extension that document more headers that SYCL applications can
include for better control of what they pay for in terms of compile
time.

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+- [code]#get_native#
+- [code]#make_*#
+
+Note that even though functions defined in the header operate with various
```
Some kind of leftover from a previous version. The intent for this note is to apply to all headers, i.e. the extension does not mandate a strict split. It is a valid (even though useless) implementation option to just include existing `<sycl/sycl.hpp>` from every new header.

Therefore, the note should probably be moved into some common section and generalized.

Also highlighting this to double-check that everyone is on board with the suggested non-strict approach.

Pennycook replied:
I agree that we should go with the non-strict approach.  I don't think there is precedent anywhere for "if you include X, I promise you won't also get Y", it would limit what implementations can do, and it would be really hard to test.  I think phrasing that says "You must include `<sycl/sycl.hpp>` or `<sycl/header>` to _guarantee_ features X, Y and Z are available" makes the most sense.

AlexeySachkov replied:
Generalized this note in ba0812d5b01c7528617baae93bddef40b6fe03df, also feedback on its wording is still welcome

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+[[sec:khr-includes-runtime]]
+=== [code]#<sycl/khr/includes/runtime># header
+
+This header groups most of classes defined by 4.6 SYCL runtime classes.
```
I would like to also highlight this one. The "most" part is very confusing, I suppose. From the compile-time point of view those classes are to small to give any significant impact being together or separate, so initial version had them together to simplify life for users. But I guess confusion doesn't means simplicity :)

You can expect that I will split those classes into separate headers in the first iteration of applying review feedback. Any comments and suggestions are always welcome, of course.

Pennycook replied:
I'm beginning to think that we should only provide the split headers, and not provide grouped headers like this.  If we only provide the split headers, we can get real-life experience with them to figure out which groupings are common, and decide later whether we want to provide them.

Note also that if we choose not to provide any grouped headers, it's trivial for a user to create their own.

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+then any macro defined by other extensions myst be made available through
+[code]#<sycl/khr/includes/version># header.
+
+== Open issues/questions
```
Immediately recording a point from @illuhad that we may want to have a header that provides forward-declarations only to gather feedback about this.

AlexeySachkov commented on adoc/chapters/architecture.adoc:
```diff
@@ -1758,6 +1758,7 @@ code as well as the <<device-image, device images>> for the
 As in the <<smcp>> case, each <<device-image>> could either contain native
 device ISA or an intermediate language.
 
+[[subsec:library-only]]
```
I can of course outline these changes into a separate PR. I needed more anchors to avoid harcoding section names and I still need a few more of them (for `vec` and `marray` aliases, in particular)

AlexeySachkov reviewed (commented):
Leaving some comments for reviewers in advance.

See [my talk at IWOCL 2025](https://youtu.be/2-NktO2VhuM) for more background/context/motivation for this extension.

As of right now, we haven't prototyped the extension ourselves in Intel's SYCL implementation, but that will be done soon after first feedback on this proposal.

CTS haven't been updated yet to have a mode which uses this extension. What I discovered is that not that many CTS tests (in fact, just one (!)) test directly includes `<sycl/sycl.hpp>` header and the rest get it through things like `common.h`. So, before CTS can use this approach and benefit from it, some refactoring has to be done (which is somewhat already in motion, see my PRs like KhronosGroup/SYCL-CTS#1083 or KhronosGroup/SYCL-CTS#1077

Proper testing for this extension could actually be tricky. For example, Intel's SYCL headers are tightly interconnected and even if you include just one of them there is a high chance that you are getting a lot more. Therefore, its hard to use them to make sure that this specification did not forget about some class or function. I had an idea about taking synopsis headers from the spec and compiling CTS with them, but even without practically doing that I see that many synopsis headers are missing and it won't be the simplest thing to do (although it is probably a good exercise to make sure that CTS does not use any undocumented APIs).

Initial version of the proposal presented here roughly provides a separate header for almost every feature, bundling some of them together (or providing a higher-level aggregation header in some cases). From compile-time savings point of view it is not necessary to make some of those headers separately (like `context`, `device`, `platform`, etc. are very cheap), but it is not immediately clear how to bundle them together for simplicity of use (or whether it is needed at all).

Note that there are already some open issues recorded at the end of the proposed document.

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+- [code]#__writeable_swizzle__# and [code]#__const_swizzle__# classes defined
+  by <<swizzled-vec-class>>
+
+Note that the simple swizzle functions ([code]#XYZW_SWIZZLE# and
+[code]#RGBA_SWIZZLE# defined by the table 123) are only available when the macro
+[code]#SYCL_SIMPLE_SWIZZLES# is defined before including
+[code]#<sycl/khr/includes/vec>#.
```
Perhaps we could add a non-normative note here saying that it is recommended for developers to define `SYCL_SIMPLE_SWIZZLES` via the command-line and not in code, since this is the only way to guarantee that the macro is defined before the first include of `vec` (which may not be in user code)?

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+
+== Open issues/questions
+
+- Should we put [code]#remove_decoration# into
+  [code]#<sycl/khr/includes/type_traits># (currently resides in
+  [code]#<sycl/khr/includes/multi_ptr>#)?
```
I think it should be in `type_traits`.  Type traits are used for generic meta-programming, and it seems to me like somebody might want to use `remove_decoration` without pulling in the whole of `multi_ptr`.

The specification says that `remove_decoration_t<T>` is `T` for undecorated types, so code like this makes sense:

```c++
#include <sycl/khr/includes/type_traits>

template <typename T>
void foo(T ptr) {
  using undecorated_t = sycl::remove_decoration_t<T>;
  bar(static_cast<undecorated_t>(ptr));
}
```

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,610 @@
+
+|[code]#<sycl/khr/includes/reduction>#
+|This header provides access to functionality related to reductions (such as
+[code]#reduction# interface and [code]#reducer class)
```
```suggestion
[code]#reduction# interface and [code]#reducer# class)
```

AlexeySachkov replied:
Applied in 2f70d45aaef845bd58f590cd75eda3dea70dc932

AlexeySachkov committed:
Fix formatting

AlexeySachkov committed:
Move `remove_decoration` into type traits header

AlexeySachkov committed:
Generalize guarantees about headers content

AlexeySachkov committed:
Split and remove runtime include

AlexeySachkov committed:
Fix section reference

AlexeySachkov committed:
Merge remote-tracking branch 'origin/main' into private/asachkov/split-sycl-hpp

AlexeySachkov commented:
Thanks for the feedback, @PeterTh!

> 1. Should `sycl.hpp` define `SYCL_KHR_INCLUDES` if the extension is supported?

Yes, I think that it is already covered by the proposal: `version.hpp` header includes extension macro and it should be included into existing `sycl/sycl.hpp`.

> 2. Do we want `submit.hpp`? It's the only instance of this pattern where the extension provides _both_ a pre-made grouping and individual headers. I have no strong opinion here.

I think that we need feedback from @Pennycook here. I'm not opposed to dropping it, neither I am opposed to having it. Both `queue` and `handler` always go side by side if accessors are in use, but with USM and in-order queues you can save on skipping `handler` as it is quite expensive. I.e. if you use `queue.submit`, you can just use `submit` header and save yourself a single include.

> 3. Should `index_space.hpp` be split up further?

My background with the split of those headers is to improve compile-times. From what I see in Intel's implementation is that those index space classes are quite small and cheap in terms of compile-time. And those classes oftentimes go together in user's applications. However, I'm looking once again at the data I gathered from [`zjin-lcf/HeCBench`](https://github.com/zjin-lcf/HeCBench) and I see that only `nd_range` and `nd_item` go together often, but not `id`, `range` and `item`:

![result2](https://github.com/user-attachments/assets/005d1308-91c1-44c7-b987-55ef0a0d3cad)

This chart shows how many benchmarks that use feature `A` (let's say rows) also used feature `B` (columns).

Though, my data gathering approach wasn't perfect, i.e. it was a plain `grep` which definitely missed all uses of `id`/`item` when a kernel lambda was defined using `auto` keyword.

I'm not opposed to splitting this header, but I do not see much value in such split.

> 4. Should `groups.hpp` be split up further?
>    This is actually the only one I feel a bit more strongly about. I think it should be possible for libraries to provide functionality which interacts with group types without pulling in all algorithms.

I can imagine that group algorithms are way heavier than any group class itself. However, what would be a good split here?
I imagine that fences and barriers are cheap and may be used more often than scans - will we have a separate header for group synchronization functions?

> 5. Should including `half.hpp` do anything for math on halfs?

The intent here was that `half.hpp` only contains `half` definition by itself, without corresponding math functions. If you need math, then you would also need to include `math.hpp`.

> 6. For `reduction.hpp`, I think the table should be clarified to note that it does not include reduction-related (sub)group stuff.

Noted, I will take a deeper look into this header, I'm not very familiar with the reductions functionality.

> 7. While making my notes, I proposed replacing `property_list.hpp` with `property.hpp` and make it also include all the properties. But I'm not really sure about that one after considering it a bit more.

I wonder if we want to have dedicated headers for different properties, because I have many questions when it comes to extensions. There could be ones which add new properties (and in fact, we have many of those at intel/llvm)
- does it mean that all those properties should be exposed through a header that corresponds to a class those properties apply to?
- what if a property is applicable to multiple classes?
- what if it isn't applicable to any class at all and instead passed to a free function (some group algorithm, for example)?

Similar questions are applicable to properties documented by the core spec, so it warrants a thorough look at them.

Pennycook commented:
> I think that we need feedback from @Pennycook here. I'm not opposed to dropping it, neither I am opposed to having it. Both `queue` and `handler` always go side by side if accessors are in use, but with USM and in-order queues you can save on skipping `handler` as it is quite expensive. I.e. if you use `queue.submit`, you can just use `submit` header and save yourself a single include.

I think we should split it up.

I can't find where I posted it, but I said something to the effect of "We should provide the most fine-grained split we're comfortable with, and let users group things if they want to."  If we _don't_ have `submit.hpp` in the first extension, people can define their own version, we can get feedback on the initial split, and potentially include it later.  If we _do_ have `submit.hpp` in the first extension, we're stuck with it forever.

> I can imagine that group algorithms are way heavier than any group class itself. However, what would be a good split here? I imagine that fences and barriers are cheap and may be used more often than scans - will we have a separate header for group synchronization functions?

What if we put the groups in `<group>` and the group algorithms in `<algorithm>`?

I think it would make sense to put `group_broadcast` and `group_barrier` in `<group>`, because they're defined as top-level group functions.  Everything else can be an opt-in where you need both `<group>` and `<algorithm>`.  (I say that because we could imagine adding some device-wide algorithms in the future, so `<algorithm>` needn't pull in `<group>` if it doesn't want to.)

AlexeySachkov committed:
Add .hpp extension to new headers

AlexeySachkov committed:
Resolve SYCL_SIMPLE_SWIZZLES issue

gmlueck commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+=== [code]#<sycl/khr/includes/atomic># header
+
+This header contains:
+- [code]#atomic_fence# function
```
Need a blank line before the bulleted list.

gmlueck commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+- [code]#memory_order_traits# struct and [code]#atomic_ref# class as defined
+  by <<sec:atomic-references>>
+- [code]#atomic# class and free functions like [code]#atomic_store#,
+  [code]#atomic_fetch_and# as defined by the section 4.15.4 Atomic types
```
Hard coded section number

AlexeySachkov replied:
Replaced with a proper reference in a55dc244a36c51cdac3627474ba28cd9ead565b5

gmlueck commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+=== [code]#<sycl/khr/includes/functional>#header
+
+This header contains:
+- [code]#plus#, [code]#multiplies#, [code]#bit_and#, [code]#bit_or#,
```
Need a blank line before the bulleted list.  In addition, should each function be its own list item?  It's weird to have a bulleted list with only one item.

AlexeySachkov replied:
The single-item list was split into multiple items in cfaf79c63c835204e2ce09b2c2a32d831d364738

gmlueck commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+=== Preprocessor directives and macros
+
+In case a SYCL implementation is made using library-only approach, macros
+defined by the section 5.6 Preprocessor directives and macros should be
```
Hard coded section number.

AlexeySachkov committed:
WIP on unifying wording

AlexeySachkov committed:
Switch to the new note syntax

AlexeySachkov committed:
Use asterics instead of dash for bullet lists

AlexeySachkov committed:
Apply some comments

AlexeySachkov committed:
Remove hardcoded section references

AlexeySachkov committed:
Document missing event info descriptors

AlexeySachkov committed:
Document missing accessor-related enumerations

AlexeySachkov committed:
Fix missing namespaces

AlexeySachkov committed:
Remove duplication in reduction.hpp description

AlexeySachkov committed:
Move identity-related type traits into the dedicated header with other type traits

AlexeySachkov committed:
Document missing kernel-device-specific info descriptors

AlexeySachkov committed:
Drop memory_order_traits because it is exposition-only

AlexeySachkov committed:
Drop submit.hpp header

AlexeySachkov committed:
Document SYCL_BACKEND_* macro

AlexeySachkov committed:
Split groups.hpp header

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+The core SYCL specification says that all APIs provided by the specification are
+accessible through single [code]#<sycl/sycl.hpp># header file.
+
+That leads to increased compilation times, because SYCL users have to pay for
```
Applied in e4233da37182a425724b4e9555d032e78e5b51c0

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+applications that aim to be portable between SYCL implementations.{endnote}
+
+If an implementation supports this extension together with other extensions,
+then any macro defined by other extensions myst be made available through
```
Applied in e4233da37182a425724b4e9555d032e78e5b51c0

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+
+All headers are provided under [code]#<sycl/khr/includes/># include path.
+
+The Description column in the table is non-formative and intended to help you
```
Applied in e4233da37182a425724b4e9555d032e78e5b51c0

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+- [code]#property_list# class definition
+
+[[sec:khr-includes-functional]]
+=== [code]#<sycl/khr/includes/functional>#header
```
Applied in e4233da37182a425724b4e9555d032e78e5b51c0

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+
+- [code]#rounding_mode# enumeration
+- [code]#elem# struct and [code]#vec# class
+- Type aliases defined by the section 4.14.2.2 Aliases
```
Replaced with a proper reference in a55dc244a36c51cdac3627474ba28cd9ead565b5

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,626 @@
+header to get [code]#accessor# definition explicitly.
+
+[NOTE]
+====
```
Switched to `{note}` syntax in bcc0d74c979938080e89590111a928a809dc80a8

AlexeySachkov commented:
> **Headers should have `.hpp` suffix.** 

Added `.hpp` suffix in 736fe3e650a3ecbb65b1f2effe81e660b588235e

> **`SYCL_SIMPLE_SWIZZLES`**. 

Recorded suggested resolution in 7e6881de691c1b75ce63aed28979b61db709795a

> **Wording**. 

Switched most of headers to use "provides" wording in 6aaa8df87736d3c35e905f2abf8243d686377006

@gmlueck, I haven't resolved all your comments yet, but most of them are addressed.

@PeterTh, @Pennycook,

- `submit.hpp` was dropped in b1daf3acc91f4800d3ee15174e3741e72a8a0cab
- `group.hpp` was split in d7cce9f955d226b33adfa2d6a302675f752f6965

Items which are still on my TODO list:
- remaining comments from @gmlueck
- `reduction.hpp` clarification from @PeterTh 
- Each property should be in its own separate header

AlexeySachkov committed:
Put kernel_handler into its own header

AlexeySachkov committed:
Merge remote-tracking branch 'origin/main' into private/asachkov/split-sycl-hpp

AlexeySachkov committed:
Split a single bullet into multiple

AlexeySachkov committed:
Outline all properties into distinct headers

AlexeySachkov committed:
Clarify the content of reduction.hpp

AlexeySachkov commented:
@PeterTh, 

I made an attempt at `reduction.hpp` clarification in 7187f28abfda609ffa95fa88fd85531dd68b10f4 - let me know if you had something else in mind.

Properties were outlined into distinct headers in 59b5e14ed8c6d020c5bda03f2d5c76cd6b484566, see also #892

With those recent commits I believe that I've addressed all feedback. Whilst I wait for other comments, I will be working on revamping the prototype I had for intel/llvm and preparing a dedicated test sub-suite in the CTS for the extension

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+many existing extensions which may not have been updated to account for this
+extension.
+
+Therefore, any extension which does not explicitly document how can it
+be accessed through header files, should be assumed to be available through
+[code]#<sycl/sycl.hpp># only.
```
```suggestion
Any extension which does not explicitly document how it can be accessed through header files
should be assumed to be available only through [code]#<sycl/sycl.hpp>#.
```

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+won't be the case and therefore this assumption is not safe to make for
+applications that aim to be portable between SYCL implementations.{endnote}
+
+If an implementation supports this extension together with other extensions,
+then any macro defined by other extensions must be made available through
+[code]#<sycl/khr/includes/version.hpp># header.
```
```suggestion
If an implementation supports this extension, then any macro defined by other
supported extensions must be defined in [code]#<sycl/khr/includes/version.hpp>#.
```

AlexeySachkov replied:
Applied in cdf04a30b74a521ec25b94a1da55798d2101b156

Pennycook commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+be accessed through header files, should be assumed to be available through
+[code]#<sycl/sycl.hpp># only.
+
+{note}It _may_ seem safe to assume that if an extension modifies some existing
+class, then its functionality should be available through a header file which
+exposes that class. However, there exist implementation possibilities where this
+won't be the case and therefore this assumption is not safe to make for
+applications that aim to be portable between SYCL implementations.{endnote}
```
```suggestion
```

AlexeySachkov committed:
Merge remote-tracking branch 'origin/main' into private/asachkov/split-sycl-hpp

AlexeySachkov committed:
Restore extra anchors lost during merge

AlexeySachkov committed:
Fix references to the core spec

AlexeySachkov committed:
Apply comments to the introduction sections of the extension

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+All headers are provided under [code]#<sycl/khr/includes/># include path.
+
+The Description column in the table is non-normative and intended to help you
+navigate the extension document quicker. Formal description of each new header
+file is provided in subsequent sections below.
```
The idea here was to give a short description which may not cover everything, like "`multi_ptr` and related address space classes" isn't exactly formal. If that's fine, then we can drop the non-normative note

AlexeySachkov committed:
Apply wording suggestions

AlexeySachkov committed:
Fix some typos

AlexeySachkov committed:
Reformat the extension overview table

AlexeySachkov committed:
Revert "Outline all properties into distinct headers"

This reverts commit 59b5e14ed8c6d020c5bda03f2d5c76cd6b484566.

AlexeySachkov committed:
Accept removal suggestions

AlexeySachkov committed:
Update code example about version.hpp

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+This extension provides an alternative set of header files which provide
+limited set of APIs each, thus allowing user to avoid paying compilation time
+cost for features that are not used by an application.
```
Applied in fe64b691ceeea0fed7625668f2592ab8cfcb1f34

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+The alternative set of headers provided by this extension does not introduce
+any new functionality, but simply provides an alternative way of accessing the
+core SYCL 2020 features with less compile-time overhead on features unused by
+an application.
```
Dropped it in fe64b691ceeea0fed7625668f2592ab8cfcb1f34

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+That leads to increased compilation times because SYCL users have to pay for
+features that they are not even using. Moreover, the situation doesn't scale
+well with more extensions being proposed.
```
Removed it in fe64b691ceeea0fed7625668f2592ab8cfcb1f34

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+|[code]#<sycl/khr/includes/index_space.hpp>#
+|This header groups together ranges and most of index space identifiers from
+<<sec:ranges-identifiers>>
```
Yeah, `sec:` is extra in this case, fixed in bba9f81996096b7952f3f25c7fd4de9d0812337c

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+[code]#joint_reduce#.
+
+|[code]#<sycl/khr/includes/stream.hpp>#
+|This header provides functionality defined by <<subsec::stream>>, i.e.
```
Fixed in bba9f81996096b7952f3f25c7fd4de9d0812337c

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+[[sec:khr-includes-full-list-of-headers]]
+== Extension overview
+
+The <<table.khr-includes-list>> below provides a list of all new
```
Applied in cdf04a30b74a521ec25b94a1da55798d2101b156

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+documented by this extension, as well as [code]#<sycl/sycl.hpp>#.
+
+There is no guarantee that the aforementioned macro are defined before the
+inclusion point of [code]#<sycl/khr/includes/version.hpp>#.
```
Applied in cdf04a30b74a521ec25b94a1da55798d2101b156

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+|[code]#<sycl/khr/includes/functional.hpp>#
+|This header provides definitions of function objects like [code]#plus# or
+[code]#locial_or# defined by <<sec:function-objects>>
```
Fixed in cde0cd224e434d352cbac3e52d8c362373abe7f6

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+
+There is no guarantee that headers documented by this extension contain anything
+else besides things listed in sections below. For example, [code]#queue# class
+has methods which accept [code]#accessor# object, but
```
Fixed in cde0cd224e434d352cbac3e52d8c362373abe7f6

AlexeySachkov committed:
Fix a couple more spec links

AlexeySachkov committed:
Definitions defined by -> definitions from

AlexeySachkov committed:
Make the overview table normative

AlexeySachkov committed:
Update comments in the code snippet

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+{note}
+Classes like [code]#h_item#, [code]#group# and [code]#sub_group# are provided by
+other headers.
+{endnote}
```
Removed the `{note}` markup and made the table normative in b7325471a84d4c3966f9b75f47e7d1edb689c587

AlexeySachkov commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,675 @@
+|This header provides defintion of the [code]#property_list# class
+
+|[code]#<sycl/khr/includes/kernel_bundle.hpp>#
+|This header provides definitions defined by <<sec:interfaces.bundles>>, such
```
"Definitions from" sounds good, I made a pass over the whole document to replace all occurrences in 4abdba609a7f2094ee9d03bd35a88a40d20f65da

KornevNikita committed:
fix typos

KornevNikita committed:
reorganize the headings

KornevNikita commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,584 @@
+
+Any extension which does not explicitly document how it can be accessed through
+header files should be assumed to be available only through
+[code]#<sycl/sycl.hpp>#.
```
@gmlueck do you have any preference here?

koparasy replied:
@gmlueck do we want to have the headers under the `sycl/khr `directory or under the `sycl/khr/includes/` directory?  

As is now we use both directories, existing khr's are under `sycl/khr` while this khr places the headers under `sycl/khr/includes`. 

Since previous `khr`'s do not have a specific independent inclusion path shouldn't we put all user includes under the same directory?

KornevNikita commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,584 @@
+{endnote}
+
+[[sec:khr-includes-version]]
+=== [code]#<sycl/khr/includes/version.hpp># header
```
I created a new section for headers - [cf2a0c8](https://github.com/KhronosGroup/SYCL-Docs/pull/814/commits/cf2a0c80e87c5e0cc6c8d6b70911e32a725e67f1). And moved the "Co-existence with other extensions" to a separate section.

KornevNikita committed:
Merge remote-tracking branch 'origin/main' into private/asachkov/split-sycl-hpp

KornevNikita committed:
Update existing KHR extensions

KornevNikita committed:
Fix pre-commit

KornevNikita commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,584 @@
+This header contains definition of [code]#byte# type alias
+
+[[sec:khr-includes-other-extensions]]
+=== Co-existence with other extensions
```
Updated - [cf2a0c8](https://github.com/KhronosGroup/SYCL-Docs/pull/814/commits/cf2a0c80e87c5e0cc6c8d6b70911e32a725e67f1)

KornevNikita committed:
Add Header files section

VerenaBeckham commented on adoc/extensions/sycl_khr_includes.adoc:
```diff
@@ -0,0 +1,592 @@
+|Description
+
+|[code]#<sycl/khr/includes/version.hpp>#
+|Provides varios macro definitions.
```
```suggestion
|Provides various macro definitions.
```

KornevNikita committed:
Update adoc/extensions/sycl_khr_includes.adoc

Co-authored-by: Verena Beckham <vbeckham@qti.qualcomm.com>

TApplencourt commented:
- Still waiting for the CTS.
-- Good proof for AI generated tests.

koparasy committed:
Update work_item_queries.hpp inclusion path for consistency

koparasy committed:
[KHR] Split builtin KHR include docs into per-section headers

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/SYCL-Docs?pullRequest=814) <br/>All committers have signed the CLA.

KornevNikita commented:
@AlexeySachkov don't know why, but it's required to sign CLA again. Could you please take a look?

koparasy committed:
Revert "Update work_item_queries.hpp inclusion path for consistency"

This reverts commit 8d0100f851dca0b17aa1f339183368ac726bae35.

koparasy committed:
[KHR] Restore math.hpp umbrella docs and narrow builtin section references

koparasy committed:
Update KHR name to split_headers

koparasy committed:
Fix formatting

koparasy committed:
[KHR] Fix split_headers extension spec inconsistencies

Apply audit fixes to sycl_khr_split_headers.adoc:

- Overview table: image.hpp filename and image_sampler symbol name now
  match per-section heading and SYCL 2020 spec.
- queue.hpp: properties::queue -> property::queue.
- event.hpp: info::event_profiling cross-reference now points to
  sec:event-profiling-info-descriptors instead of sec:event-info-descriptors.
- image.hpp: fix addresing_mode typo and add image_target enum (required
  template parameter of image accessor classes).
- kernel_bundle.hpp: enumerate the full set of free functions defined by
  the core spec (build, join, is_compatible, get_kernel_id, get_kernel_ids)
  instead of the open-ended "like ..." list.
- vec.hpp / marray.hpp: swap the cross-references to subsec:vec-aliases
  and subsec:marray-aliases that were previously inverted.
- group_algorithms.hpp: replace shorthand names (any_of, reduce, ...) with
  the actual SYCL 2020 function names (any_of_group / joint_any_of,
  reduce_over_group / joint_reduce, shift_group_left, shift_group_right,
  permute_group_by_xor, select_from_group). All 17 functions enumerated.
- group_algorithms.hpp: drop doubled "functions functions".

KornevNikita commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+supported extensions must be defined in
+[code]#<sycl/khr/split_headers/version.hpp>#.
+
+== Open issues/questions
```
IIUC the SYCL-CTS patch has been merged, so this patch should also be close to the finish line. Previously @gmlueck suggested to remove this section in case if there are no issues/questions, but I've postponed this until this patch is ready for merge.

gmlueck replied:
We plan to have a formal vote on Thursday for this KHR, so it is time to remove this section.

koparasy replied:
Removed in f93fb84

gmlueck commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+
+In order to make simple swizzle functions ([code]#XYZW_SWIZZLE# and
+[code]#RGBA_SWIZZLE# from <<table.members.vec>>) available, the macro
+[code]#SYCL_SIMPLE_SWIZZLES# should be defined before *any* other [code]#sycl/#
```
```suggestion
[code]#SYCL_SIMPLE_SWIZZLES# must be defined before *any* other [code]#sycl/#
```

gmlueck commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+[[sec:khr-split-headers-byte]]
+=== [code]#<sycl/khr/split_headers/byte.hpp># header
+
+This header contains definition of [code]#byte# type alias
```
```suggestion
This header contains the definition of the [code]#byte# type alias
```

gmlueck commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+
+[[sec:khr-split-headers-hierarchical-parallelism]]
+=== [code]#<sycl/khr/split_headers/hierarchical_parallelism.hpp># header
+
```
I think we should add:

> Deprecated by SYCL 2020

to indicate that the header itself is deprecated in that SYCL version.

koparasy replied:
Done in 747b91f

gmlueck commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+
+[[sec:khr-split-headers-byte]]
+=== [code]#<sycl/khr/split_headers/byte.hpp># header
+
```
> Deprecated by SYCL 2020

Because the `byte` alias is deprecated.

koparasy replied:
Done in 747b91f

gmlueck commented on adoc/extensions/sycl_khr_split_headers.adoc:
```diff
@@ -0,0 +1,648 @@
+
+The Description column in the table is intended to help you navigate the
+extension document quicker.
+It does not offer the exaustive listing of each new header file contents that
```
```suggestion
It does not offer the exhaustive listing of each new header file contents that
```

gmlueck reviewed (commented):
A few last-minute edits.  Mostly spelling and grammar, but:

* Change one occurrence of "should" to "must"
* Add two occurrences of "Deprecated by SYCL 2020" to indicate that two of the headers in the KHR are deprecated.  In both cases, all the identifiers provided by the headers are deprecated in SYCL 2020, so I presume we want to deprecate the headers also.

koparasy committed:
Apply suggestions from code review

Co-authored-by: Greg Lueck <gregory.m.lueck@intel.com>

koparasy committed:
Remove the  section 'Open Issues/questions'

koparasy committed:
Add deprecation
